Define Multi-Application Bindings spec extension (#55)
k8s-service-bindings/spec: README.md

  binding:
    name: kafka-eftg9
```

## Multi-Application Bindings

The `ServiceBinding` resource is limited by the [Service Binding](#service-binding) spec to bind a single service by-name to a single application by-name. If the application's name is generated and not known in advance, the binding implicitly must be created after the application. While level-based reconciliation will converge on the desired state, the application will roll out initially without the bound service. Depending on the application this may cause a crash loop or a misconfiguration (e.g. using an in-memory database designed for development in a production environment). By switching the application reference from using a name to a label selector the service can be bound to the application resource as it is created.

A `ServiceBinding` **MAY** define the application reference by-name or by-[label selector][ls]. A name and selector **MUST NOT** be defined in the same reference.

> As label selectors are inherently queries that return zero-to-many resources, implementations **SHOULD** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **SHOULD** be aggregated and reported on the binding status's `Ready` condition.

Author: ```suggestion
As label selectors are inherently queries that return zero-to-many resources, implementations **MUST** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **MUST** be aggregated and reported on the binding status's `Ready` condition.
```

There was a lot of discussion on the issue about allowing for an implementation to only reconcile a single application that matches the selector. The original language is allowing for that case without encouraging the behavior. My preference would be to apply this recommendation and force implementors of this extension to properly handle multiple applications matching the selector.

Reviewer: ya - I am still on the side of a single cardinality label, even if that goes against its natural usage.  :) 

The simplicity and ease of implementation of a 1-to-1 mapping between service and app, in my opinion, is a very strong aspect of the spec.    I think the sample you have chosen is a good one as it demonstrates a pattern to make a the singleton target easier:  the more labels used, the easier it is to "pick" a single instance.

Author: @arthurdm do you have concerns with the current language if we drop this suggestion?

Reviewer: My only concern is the current language steers an implementation (and user) towards multi-cardinality.  Perhaps it could be augmented to provide guidance on both options - the "missing" one being guidance on using unique label combinations (1..N) to target a single app.

Author: Single application cardinality will lead to [user surprise](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) when creating a second application that matches the selector. For that reason, it is behavior that shouldn't be encouraged by the spec. From my experience implementing this behavior, it would have been more work to restrict the selector to a single matching resource since it is introducing partial failure as a desired state. Even if that wasn't my experience, since this is an optional extension ease of implementation is less of a concern over a good DX

I respect that an implementation may have its reasons for not supporting multiple applications. Which is why I wrote the PR text the way it is as a nudge to implementors to do the user-friendly behavior without a mandate.

@arthurdm would you like to propose language that addresses the concern?

Reviewer: ya - it's a tough one.   Given that restricting a label to a single matching resource goes against the expected behaviour, I am stepping back and questioning if using a label is the way to solve this problem.  

How common is the scenario where we don't know the name of a resource, but know its labels?   Could the `ServiceBinding` CR creation be deferred until the system knows the generated name of the app resource (but still before the app is up and running)?

Author: > Could the ServiceBinding CR creation be deferred until the system knows the generated name of the app resource (but still before the app is up and running)?

The user would need to:
- create an invalid application resource (anything that blocks pod scheduling, but is structurally valid)
- get the name of the application resource
- create the `ServiceBinding` with the generated application name
- update the application to be valid

Workable, but not a friendly solution.

> How common is the scenario where we don't know the name of a resource, but know its labels?

Generated resource names are more common on controlled resources than uncontrolled. If you had a high level resource that was not PodSpec-able, but reconciled to a resource that was PodSpec-able, the service injection would need to occur on the controlled resource. If that controlled resource also uses a generated name, the only way to effectively target it and all future child resources is via a label selector.

> I am stepping back and questioning if using a label is the way to solve this problem.

It may be that a label selector isn't appropriate for the use case you originally envisioned, but I think there is still a good bit of value in supporting service binding via a selector.

Author: Resolving in favor of https://github.com/application-stacks/service-binding-specification/pull/55#discussion_r443903399

Author: Moving over to https://github.com/application-stacks/service-binding-specification/pull/55#discussion_r444324590

> As label selectors are inherently queries that return zero-to-many resources, implementations **SHOULD** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **SHOULD** be aggregated and reported on the binding status's `Ready` condition.

Reviewer: ```suggestion
As label selectors are inherently queries that return zero-to-many resources, it is **RECOMMENDED* that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors of this extension **SHOULD** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **SHOULD** be aggregated and reported on the binding status's `Ready` condition.
```

Suggesting an update that presents both worlds (single and multi resources).

> As label selectors are inherently queries that return zero-to-many resources, it is **RECOMMENDED* that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors of this extension **SHOULD** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **SHOULD** be aggregated and reported on the binding status's `Ready` condition.

Reviewer: ```suggestion
As label selectors are inherently queries that return zero-to-many resources, it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors of this extension **SHOULD** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **SHOULD** be aggregated and reported on the binding status's `Ready` condition.
```

> As label selectors are inherently queries that return zero-to-many resources, it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors of this extension **SHOULD** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **SHOULD** be aggregated and reported on the binding status's `Ready` condition.

Author: ```suggestion
As label selectors are inherently queries that return zero-to-many resources, it is **RECOMMENDED** that `ServiceBinding` authors use a combination of labels that yield a single resource, but implementors of this extension **MUST** handle each matching resource as if it was specified by name in a distinct `ServiceBinding` resource. Partial failures **MUST** be aggregated and reported on the binding status's `Ready` condition.
```

Reopening https://github.com/application-stacks/service-binding-specification/pull/55#discussion_r442924014 based on discussion on the interlock call

[ls]: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors

### Resource Type Schema

```yaml

> apiVersion: service.binding/v1alpha1

Author: ```suggestion
apiVersion: service.binding/{vendored-version}
```

kind: ServiceBinding
metadata:
  ...
spec:
  application:          # PodSpec-able resource ObjectReference-able
    ...
    name:               # string, optional, mutually exclusive with selector
    selector:           # metav1.LabelSelector, optional, mutually exclusive with name
  ...
status:
  ...
```

### Example Resource

```yaml

> apiVersion: service.binding/v1alpha1

Author: ```suggestion
apiVersion: service.binding/v1alpha1-vendor
```

kind: ServiceBinding
metadata:
  name: online-banking-to-account-service
spec:
  name: account-service

  application:
    apiVersion: apps/v1
    kind:       Deployment

>     selector:

Reviewer: I like this sample - helps to visualize.    One thing in my mind is the conversation from #52 and whether we'll put all of the extensions in a single place or keep them in-line with the rest of the core spec.

Author: yea, I've been torn on that question. The `apiVersion` should be something custom until we make the decision to inline extensions into the standard resource structure.

Author: moved the example versions away from v1alpha1
